File: interfaces/cython/cantera/examples/thermo/isentropic.py
"""
Isentropic, adiabatic flow example - calculate area ratio vs. Mach number curve

Requires: cantera >= 2.5.0, matplotlib >= 2.0
"""

import cantera as ct
import math
import numpy as np


def soundspeed(gas):
    """The speed of sound. Assumes an ideal gas."""

    gamma = gas.cp / gas.cv
    return math.sqrt(gamma * ct.gas_constant
                     * gas.T / gas.mean_molecular_weight)


def isentropic(gas=None):
    """
    In this example, the area ratio vs. Mach number curve is computed. If a gas
    object is supplied, it will be used for the calculations, with the
    stagnation state given by the input gas state. Otherwise, the calculations
    will be done for a 10:1 hydrogen/nitrogen mixture with stagnation T0 = 1200
    K, P0 = 10 atm.
    """
    if gas is None:
        gas = ct.Solution('gri30.yaml')
        gas.TPX = 1200.0, 10.0*ct.one_atm, 'H2:1,N2:0.1'

    # get the stagnation state parameters
    s0 = gas.s
    h0 = gas.h
    p0 = gas.P

    mdot = 1  # arbitrary
    amin = 1.e14

    data = np.zeros((200, 4))

    # compute values for a range of pressure ratios
    for r in range(200):

        p = p0*(r+1)/201.0
        # set the state using (p,s0)
        gas.SP = s0, p

        v2 = 2.0*(h0 - gas.h)      # h + V^2/2 = h0
        v = math.sqrt(v2)
        area = mdot/(gas.density*v)    # rho*v*A = constant
        amin = min(amin, area)
        data[r, :] = [area, v/soundspeed(gas), gas.T, p/p0]

    data[:, 0] /= amin

    return data


if __name__ == "__main__":
    print(__doc__)
    data = isentropic()
    try:
        import matplotlib.pyplot as plt
        plt.plot(data[:, 1], data[:, 0])
        plt.ylabel('Area Ratio')
        plt.xlabel('Mach Number')
        plt.title('Isentropic Flow: Area Ratio vs. Mach Number')
        plt.show()

    except ImportError:
        print('area ratio,   Mach number,   temperature,   pressure ratio')
        print(data)

File: samples/matlab_experimental/isentropic.m
function isentropic(g)
    %% Isentropic, adiabatic flow
    %
    % In this example, the area ratio vs. Mach number curve is computed for a
    % hydrogen/nitrogen gas mixture.
    %
    % .. tags:: Matlab, thermodynamics, compressible flow, plotting

    clear all
    close all

    tic
    help isentropic

    if nargin == 1
        gas = g;
    else
        gas = Solution('gri30.yaml', 'gri30');
    end

    % set the stagnation state
    gas.TPX = {1200.0, 10.0 * OneAtm, 'H2:1,N2:0.1'};
    gas.basis = 'mass';
    s0 = gas.S;
    h0 = gas.H;
    p0 = gas.P;

    mdot = 1; % arbitrary

    mach = [];
    a = [];
    i = 1;
    amin = 1.e14;

    % compute values for a range of pressure ratios
    for r = 0.005:0.0025:0.995
        p = p0 * r;

        % set the state using (p,s0)
        gas.SP = {s0, p};

        h = gas.H;
        rho = gas.D;

        v2 = 2.0 * (h0 - h); %   h + V^2/2 = h0
        v = sqrt(v2);
        a(i) = mdot / (rho * v); %   rho*v*A = constant

        if a(i) < amin
            amin = a(i);
        end

        mach(i) = v / gas.soundSpeed;
        i = i + 1;
    end

    a = a / amin;

    % plot results

    clf;
    plot(mach, a);
    ylabel('Area Ratio');
    xlabel('Mach Number');
    title('Isentropic Flow: Area Ratio vs. Mach Number');

    toc
end
